### PR TITLE
Change LifecycleListener

### PR DESCRIPTION
**Pull request by yemreinci**
Applies changes to `LifecycleListener`, similar to those in #604 and #607. Also introduces an alternative for `boost::latch` used in examples.
To summarize:
- `LifecycleListener` is now a concrete type, not an interface.
- `HazelcastClient::addLifecycleListener` and `LifecycleService::addLifecycleListener` now take the `LifecycleListener` as an rvalue reference and return a randomly generated boost::uuids::uuid as the registration id. 
- Instances of `boost::latch` are replaced with pairs of `std::promise<void>` and `std::future<void>` in examples `StartAsync.cpp`, `DoNotReconnectToCluster.cpp`, `ReconnectAsync.cpp`. The same method can be applied to other examples as well.

### CODE REVIEW

**Comment by yemreinci** [comment]
Windows build failed at first because bcrypt.lib (required by Boost.UUID) wasn't auto-linked. Related issues: https://github.com/boostorg/uuid/issues/68, https://github.com/microsoft/vcpkg/issues/4481 .
Defining `BOOST_UUID_FORCE_AUTO_LINK` solved the problem.

File: CMakeLists.txt [hunk 1]
```diff
@@ -195,7 +195,7 @@ ENDIF(${CMAKE_SYSTEM_NAME} MATCHES "Linux")
 
 IF(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
 	message(STATUS "WINDOWS ENVIRONMENT DETECTED ${CMAKE_GENERATOR} ${CMAKE_BUILD_TYPE} ")
-	add_definitions(-D__WIN${HZ_BIT}__ -DWIN${HZ_BIT} -D_WIN${HZ_BIT})
+	add_definitions(-D__WIN${HZ_BIT}__ -DWIN${HZ_BIT} -D_WIN${HZ_BIT} -DBOOST_UUID_FORCE_AUTO_LINK)
```
Comment (ihsandemir) [inline CMakeLists.txt:198]: will windows users have to compile their application with this flag as well? lets discuss alternatives.

File: hazelcast/src/hazelcast/client/spi.cpp [hunk 2]
```diff
@@ -438,10 +444,10 @@ namespace hazelcast {
                         break;
                 }
 
-                for (std::unordered_set<LifecycleListener *>::iterator it = listeners.begin(); it != listeners.end(); ++it) {
-                    (*it)->stateChanged(lifecycleEvent);
+                for (auto &item: listeners) {
+                    auto& listener = item.second;
+                    listener.stateChanged(lifecycleEvent);
```
Comment (ihsandemir) [inline hazelcast/src/hazelcast/client/spi.cpp:449]: should we consider making the interface a different function pointer line on_connected, on_shutdown, etc. ?

File: hazelcast/include/hazelcast/client/impl/HazelcastClientInstanceImpl.h [hunk 3]
```diff
@@ -171,14 +172,14 @@ namespace hazelcast {
                 *
                 * @param lifecycleListener Listener object
                 */
-                void addLifecycleListener(LifecycleListener *lifecycleListener);
+                boost::uuids::uuid addLifecycleListener(LifecycleListener &&lifecycleListener);
```
Comment (ihsandemir) [inline hazelcast/include/hazelcast/client/impl/HazelcastClientInstanceImpl.h:175]: Can we change this API to work similar to Java APi where we do `getLifecycleListener().addLifeCycleListener()` ? We were not able to change this before but now we can break it and change. LifecycleService also provides a few more public APIs such as `isRunning`. We can also provide this as public, they are currently on the client itself. This can be on another PR.

File: hazelcast/src/hazelcast/client/spi.cpp [hunk 4]
```diff
@@ -339,29 +345,59 @@ namespace hazelcast {
...
+                            listener.starting();
+                        }
                         break;
                     }
-                    case LifecycleEvent::STARTED :
+                    case LifecycleEvent::STARTED : {
                         logger.info("LifecycleService::LifecycleEvent STARTED");
+
+                        for (auto &item: listeners) {
+                            auto& listener = item.second;
```
Comment (ihsandemir) [inline hazelcast/src/hazelcast/client/spi.cpp:359]: this code looks like repeating, can you refactor, and pass the function pointer?

**Review by ihsandemir — CHANGES_REQUESTED** [review]

File: hazelcast/include/hazelcast/client/impl/HazelcastClientInstanceImpl.h (hunk 3 above)
Comment (yemreinci) [reply hazelcast/include/hazelcast/client/impl/HazelcastClientInstanceImpl.h:175]: Will do on a separate PR.

File: hazelcast/src/hazelcast/client/spi.cpp (hunk 4 above)
Comment (yemreinci) [reply hazelcast/src/hazelcast/client/spi.cpp]: done.

File: hazelcast/src/hazelcast/client/spi.cpp [hunk 5]
```diff
@@ -346,57 +348,56 @@ namespace hazelcast {
...
-                        }
+                        };
                         break;
                     }
                     case LifecycleEvent::STARTED : {
                         logger.info("LifecycleService::LifecycleEvent STARTED");
 
-                        for (auto &item: listeners) {
-                            auto& listener = item.second;
+                        fire_one = [](LifecycleListener &listener) {
```
Comment (ihsandemir) [inline hazelcast/src/hazelcast/client/spi.cpp:359]: line format - indentation

**Review by ihsandemir — APPROVED** [review]

File: hazelcast/src/hazelcast/client/spi.cpp [hunk 6]
```diff
@@ -339,29 +347,58 @@ namespace hazelcast {
...
+                            listener.disconnected();
+                        };
                         break;
-                }
+                    }
 
-                for (std::unordered_set<LifecycleListener *>::iterator it = listeners.begin(); it != listeners.end(); ++it) {
-                    (*it)->stateChanged(lifecycleEvent);
+                    for (auto &item: listeners) {
```
Comment (ihsandemir) [inline hazelcast/src/hazelcast/client/spi.cpp:397]: loop outside switch

**Review by ihsandemir — CHANGES_REQUESTED** [review]

**Comment by yemreinci** [comment]
verify-linux

**Comment by ihsandemir** [comment]
verify-linux

**Review by ihsandemir — APPROVED** [review]
